### PR TITLE
Login and delete comment features

### PR DESCRIPTION
**Pull request by lilian-hunt**

### CODE REVIEW

File: portfolio/bin/pom.xml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    <!-- This project uses Java 8 -->
+    <maven.compiler.source>1.8</maven.compiler.source>
+    <maven.compiler.target>1.8</maven.compiler.target>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <failOnMissingWebXml>false</failOnMissingWebXml>
+  </properties>
+
+  <dependencies>
+    <dependency>
+        <groupId>javax.servlet</groupId>
```
Comment (KathleenQ) [inline portfolio/bin/pom.xml]: I think you need to change the 4 space to 2 space in the "dependency" section here.

File: portfolio/bin/src/main/webapp/index.html [hunk 2]
```diff
@@ -0,0 +1,73 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <meta charset="UTF-8">
+        <title>Lilian Hunt</title>
```
Comment (KathleenQ) [inline portfolio/bin/src/main/webapp/index.html]: Still the 4-space problem throughout this file.

File: portfolio/bin/src/main/webapp/projects.html [hunk 3]
```diff
@@ -0,0 +1,32 @@
+
+<!DOCTYPE html>
+<html>
+    <head>
+        <meta charset="UTF-8">
+        <title>Lilian Hunt</title>
```
Comment (KathleenQ) [inline portfolio/bin/src/main/webapp/projects.html]: Please fix the 4-space problem throughout this file.

File: portfolio/bin/src/main/webapp/style.css [hunk 4]
```diff
@@ -0,0 +1,26 @@
+#content {
+    margin-left: auto;
```
Comment (KathleenQ) [inline portfolio/bin/src/main/webapp/style.css]: Please also change the beginning space from 4 to 2 for this file.

File: portfolio/bin/src/main/webapp/style.css [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+#fact-container, #json-msg-container{
+    margin-top: 10px;
+}
+
+img {
+    max-width: 100%;
+    max-height: 100%;
+    /* object-fit: cover; */
+    display: block; /* remove extra space below image */
```
Comment (aannie-y) [inline portfolio/bin/src/main/webapp/style.css]: If you don't need the comment there, remove them to declutter.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -43,10 +48,13 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    HashMap<String, ArrayList<String>> comments = new HashMap<String, ArrayList<String>>();
     for (Entity entity : results.asIterable()) {
-      String comment = (String) entity.getProperty("comment");
-      comments.add(comment);
+      String id = KeyFactory.keyToString(entity.getKey());
+      ArrayList<String> commentEmail = new ArrayList<String>();
+      commentEmail.add((String) entity.getProperty("comment"));
+      commentEmail.add((String) entity.getProperty("userEmail"));
+      comments.put(id, commentEmail);
     }
```
Comment (aannie-y) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Indentation should be 2 spaces as per [here](https://google.github.io/styleguide/javaguide.html#s4.2-block-indentation)

**Review by aannie-y — COMMENTED** [review]
A lot of the files will need indentation fixed from 4 to 2 spaces. There is also some comments that can be removed.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 6 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This indentation is 2 spaces

File: portfolio/bin/src/main/webapp/index.html (hunk 2 above)
Comment (lilian-hunt) [reply portfolio/bin/src/main/webapp/index.html]: fixed

File: portfolio/bin/src/main/webapp/projects.html (hunk 3 above)
Comment (lilian-hunt) [reply portfolio/bin/src/main/webapp/projects.html]: fixed

File: portfolio/bin/pom.xml (hunk 1 above)
Comment (lilian-hunt) [reply portfolio/bin/pom.xml]: fixed

File: portfolio/bin/src/main/webapp/style.css (hunk 4 above)
Comment (lilian-hunt) [reply portfolio/bin/src/main/webapp/style.css]: fixed

File: portfolio/bin/src/main/webapp/style.css (hunk 5 above)
Comment (lilian-hunt) [reply portfolio/bin/src/main/webapp/style.css]: removed, was removed before in the last pull request, not sure why it didn't update with git pull origin master

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -56,11 +64,10 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   /*
    * Helper function to store an array in JSON format
```
Comment (tripd22) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Nit: update comment (converting a Map now, not a List)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -71,8 +78,10 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     // Get the input from the form.
     String feedback = request.getParameter("text-input");
     long timestamp = System.currentTimeMillis();
+    UserService userService = UserServiceFactory.getUserService();
```
Comment (tripd22) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: It might be better to store this UserService as a private instance variable so we don't have to create a new one every time doPost() is called.

You could also store the DatastoreService in a similar way, because at the moment both doPost() and doGet() use it.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 6 above)
Comment (jupi3200) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Did you run clang-format?

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    Key key = KeyFactory.stringToKey(request.getParameter("id"));
+    String commentUserEmail = request.getParameter("userEmail");
+
+    // Get the email of the user currently logged in.
+    UserService userService = UserServiceFactory.getUserService();
+    String currentUserEmail = userService.getCurrentUser().getEmail();
+
+    // Users can only delete their own comments.
+    if (!commentUserEmail.equals(currentUserEmail)) {
+      System.out.println("Cannot delete other user's comments!");
```
Comment (jupi3200) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: Could we use log?

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      response.sendRedirect("/index.html");
```
Comment (jupi3200) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: Do we expect the web page to allow users to send such requests?
Should we inform the user their request failed due to ... ?

**Comment by jupi3200** [comment]
Can this PR be split to smaller parts (for example login and delete need not be in the same request).

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 6 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Yes I did

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 11]
```diff
@@ -43,10 +48,13 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
-    List<String> comments = new ArrayList<String>();
+    HashMap<String, ArrayList<String>> comments = new HashMap<String, ArrayList<String>>();
     for (Entity entity : results.asIterable()) {
-      String comment = (String) entity.getProperty("comment");
-      comments.add(comment);
+      String id = KeyFactory.keyToString(entity.getKey());
+      ArrayList<String> commentEmail = new ArrayList<String>();
```
Comment (tripd22) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: We typically use the interface for the type of collection variables, i.e.

List<String> commentEmail = new ArrayList<String>();   // uses interface as type
ArrayList<String> commentEmail = new ArrayList<String>();    // uses class as type

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 12]
```diff
@@ -43,10 +48,13 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
-    List<String> comments = new ArrayList<String>();
+    HashMap<String, ArrayList<String>> comments = new HashMap<String, ArrayList<String>>();
```
Comment (jupi3200) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Would _new HashMap<>();_  be sufficient?

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java (hunk 10 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: Now creates a pop up

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 7 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: changed

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 12 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: yep, updated

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 11 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: changed this one and also when creating the Hashmap

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: implemented private variable for both

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 13]
```diff
@@ -17,50 +17,57 @@
...
 import com.google.appengine.api.datastore.PreparedQuery;
 import com.google.appengine.api.datastore.Query;
 import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
 import com.google.gson.Gson;
 import java.io.IOException;
-import java.util.ArrayList;
 import java.util.List;
+import java.util.ArrayList;
```
Comment (jupi3200) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:28]: https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 14]
```diff
@@ -17,50 +17,57 @@
...
-    List<String> comments = new ArrayList<String>();
+    Map<String, List<String>> comments = new HashMap<>();
     for (Entity entity : results.asIterable()) {
-      String comment = (String) entity.getProperty("comment");
-      comments.add(comment);
+      String id = KeyFactory.keyToString(entity.getKey());
+      List<String> commentEmail = new ArrayList<>();
+      commentEmail.add((String) entity.getProperty("comment"));
+      commentEmail.add((String) entity.getProperty("userEmail"));
+      comments.put(id, commentEmail);
```
Comment (jupi3200) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:57]: Why are we placing different types of attributes in a string list?
This forces the client to depend on the ordering of elements in a list, and makes the code fragile to change.
What alternative representation would be better?

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
+// Copyright 2019 Google LLC
```
Comment (jupi3200) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:1]: Should this be in a different PR?

**Comment by jupi3200** [comment]
Please keep PRs shorter.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 14 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:57]: I am now using a map to store the information.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 13 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:28]: Run clang-format to fix this, sorry I didn't run it immediately before submitting my code

**Comment by lilian-hunt** [comment]
Sorry yes I am trying to keep my PRs short but the last PR took so long to approve that all my work built up so once this PR is done, I will be submitting small regular PRs. Also the changes to the HTML and CSS files are just the spacing so it might look like a lot of changes have been made but not really in fact.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 12 above)
Comment (jupi3200) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: The code still looks the same.

Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Screen Shot 2020-12-21 at 1 07 18 pm
I have no idea why. This is my code in the shell editor. I have tried to push the updates but it says that everything is up to date.

Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I will ask tech support now

Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: ok so when you click on view file you need to make sure the right branch/latest commit is selected to see all the changes

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java (hunk 9 above)
Comment (lilian-hunt) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: Deleted the print statement, not really needed, it now notifies the user directly

**Comment by lilian-hunt** [comment]
All comments have been addressed now

**Review by jupi3200 — APPROVED** [review]
